tests/unit/s2n_fork_generation_number_test.c

```diff
@@ -0,0 +1,349 @@
+    EXPECT_EQUAL(return_fork_generation_number, 0);
+
+    /* Should be idempotent in threaded environment as well */
+    EXPECT_EQUAL(s2n_unit_test_thread(return_fork_generation_number), S2N_SUCCESS);
+
+    /* Cached FGN should increment if a fork event occurs */
+    EXPECT_EQUAL(s2n_unit_test_fork(return_fork_generation_number, FORK_LEVEL_FOR_TESTS), S2N_SUCCESS);
+    EXPECT_EQUAL(s2n_unit_test_fork_check_threads_first(return_fork_generation_number), S2N_SUCCESS);
+
+    /* Some fork detection mechanisms can also detect forks through clone() */
+    if (test_case->test_case_must_pass_clone_test == CLONE_TEST_YES) {
+        EXPECT_EQUAL(s2n_assert_madv_wipeonfork_is_supported() || s2n_assert_map_inherit_zero_is_supported(), true);
+        EXPECT_EQUAL(s2n_unit_test_clone(return_fork_generation_number), S2N_SUCCESS);
+    }
+    else if (test_case->test_case_must_pass_clone_test == CLONE_TEST_DETERMINE_AT_RUNTIME) {
+        if (s2n_assert_madv_wipeonfork_is_supported() == true ||
+            s2n_assert_map_inherit_zero_is_supported() == true) {
+            EXPECT_EQUAL(s2n_unit_test_clone(return_fork_generation_number), S2N_SUCCESS);
+        }
+    }
+
+    return S2N_SUCCESS;
+}
+
+static int s2n_test_case_default_cb(struct fgn_test_case *test_case)
+{
+    EXPECT_EQUAL(s2n_unit_tests_common(test_case), S2N_SUCCESS);
+
+    return S2N_SUCCESS;
+}
+
+static int s2n_test_case_pthread_atfork_cb(struct fgn_test_case *test_case)
+{
+    POSIX_GUARD_RESULT(s2n_ignore_wipeonfork_and_inherit_zero_for_testing());
+    EXPECT_EQUAL(s2n_unit_tests_common(test_case), S2N_SUCCESS);
+
+    return S2N_SUCCESS;
+}
+
+static int s2n_test_case_madv_wipeonfork_cb(struct fgn_test_case *test_case)
+{
+    if (s2n_assert_madv_wipeonfork_is_supported() == false) {
+        FGN_TEST_CASE_PRINT_MSG_INFO("Test case not supported. Skipping.", test_case);
+        return S2N_SUCCESS;
+    }
+
+    POSIX_GUARD_RESULT(s2n_ignore_pthread_atfork_for_testing());
+    EXPECT_EQUAL(s2n_unit_tests_common(test_case), S2N_SUCCESS);
+
+    return S2N_SUCCESS;
+}
+
+static int s2n_test_case_map_inherit_zero_cb(struct fgn_test_case *test_case)
+{
+    if (s2n_assert_map_inherit_zero_is_supported() == false) {
+        FGN_TEST_CASE_PRINT_MSG_INFO("Test case not supported. Skipping.", test_case);
+        return S2N_SUCCESS;
+    }
+
+    POSIX_GUARD_RESULT(s2n_ignore_pthread_atfork_for_testing());
+    EXPECT_EQUAL(s2n_unit_tests_common(test_case), S2N_SUCCESS);
+
+    return S2N_SUCCESS;
+}
+
+struct fgn_test_case fgn_test_cases[NUMBER_OF_FGN_TEST_CASES] = {
+    {"Default fork detect mechanisms.", s2n_test_case_default_cb, CLONE_TEST_DETERMINE_AT_RUNTIME},
+    {"Only pthread_atfork fork detection mechanism.", s2n_test_case_pthread_atfork_cb, CLONE_TEST_NO},
+    {"Only madv_wipeonfork fork detection mechanism.", s2n_test_case_madv_wipeonfork_cb, CLONE_TEST_YES},
+    {"Only map_inheret_zero fork detection mechanism.", s2n_test_case_map_inherit_zero_cb, CLONE_TEST_YES}
+};
```

Comment:
Any way to double-check that each of these is being run at least once in our CI? Like can you assert that specific environments are running specific tests, so that we know at least one environment is running them?

Reply:
I thought about this. there are a few options:
* If the `.yml` file running one of the CI dimensions is aware of the platform it is running on, we can somehow inject that knowledge into the unit test. 
* The unit test interrogates the system at run-time and acquire the knowledge.

There are different methods to inject the knowledge into the unit test for the former first option above:
* Use a build flag
* Set an environment variable that can be read at run-time

I vote for using `.yml` files (this gives the strongest assertion imo) and  use an environment variable to inform the unit test at run-time.

Any preferences?

Reply:
Passing information in seems like a pretty heavy-weight solution. Is there any reason interrogating the system at run-time wouldn't work?

Reply:
This seems like it's probably worth creating a separate issue, not blocking this PR

Reply:
Will create when this PR resolves - I promise :)

Reply:
https://github.com/aws/s2n-tls/issues/3240